rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,853 @@
+                self.backend_secret_key = backend_info.1;
+                None
+            }
+            PgsqlBEMessage::ReadyForQuery(_) => Some(PgsqlStateProgress::ReadyForQueryReceived),
+            // TODO should we store any Parameter Status in PgsqlState?
+            PgsqlBEMessage::AuthenticationMD5Password(_)
+            | PgsqlBEMessage::AuthenticationCleartextPassword(_) => {
+                Some(PgsqlStateProgress::SimpleAuthenticationReceived)
+            }
+            PgsqlBEMessage::RowDescription(_) => Some(PgsqlStateProgress::DataRowReceived),
+            PgsqlBEMessage::CommandComplete(_) => {
+                // TODO Do we want to compare the command that was stored when
+                // query was sent with what we received here?
+                Some(PgsqlStateProgress::CommandCompletedReceived)
+            }
+            PgsqlBEMessage::ErrorResponse(_) => Some(PgsqlStateProgress::ErrorMessageReceived),
+            _ => {
+                // TODO handle unexpected situations here?
+                // We don't always have to change current state when we see a response...
+                None
+            }
+        }
+    }
+
+    fn state_based_resp_parsing(
+        state: PgsqlStateProgress, input: &[u8],
+    ) -> Result<(&[u8], parser::PgsqlBEMessage), nom::Err<(&[u8], nom::error::ErrorKind)>> {
+        if state == PgsqlStateProgress::SSLRequestReceived {
+            parser::parse_ssl_response(input)
+        } else {
+            parser::pgsql_parse_response(input)
+        }
+    }
+
+    fn parse_response(&mut self, input: &[u8], flow: *const Flow) -> AppLayerResult {
+        // We're not interested in empty responses.
+        if input.len() == 0 {
+            return AppLayerResult::ok();
+        }
+
+        if self.response_gap {
+            if !probe_tc(input) {
+                // Out of sync, we'll just try again next time.
+                SCLogDebug!("Suricata interprets there's a gap in the response");
+                return AppLayerResult::ok();
+            }
+
+            // It seems we're in sync with a message header, clear gap state and keep parsing.
+            self.response_gap = false;
+        }
+
+        let mut start = input;
+        while start.len() > 0 {
+            match PgsqlState::state_based_resp_parsing(self.state_progress, start) {
+                Ok((rem, response)) => {
+                    start = rem;
+                    SCLogDebug!("Response is {:?}", &response);
+                    if let Some(state) = self.response_get_next_state(&response, flow) {
+                        self.state_progress = state;
+                    };
+                    if let Some(tx) = self.find_or_create_tx() {
+                        tx.responses.push(response);
+                    } else {
+                        // Not sure if this is the best solution here, almost sure it isn't...
+                        return AppLayerResult::err();
```

Comment:
hmm I don't know about the proto but generally, we go about as the function name suggests `find_or_create_tx`, if it's found, push to its `responses`, if not, create a new tx and push to its `responses`.

Reply:
the problem was that with the current implementation there could be some odd case where the `last_mut()` could panic when unwrapped, because we don't always created a new transaction (as it's StateProgress dependent). This was the solution I came up with, but it indeed doesn't look great...

Reply:
So, am I correct in understanding that when `last_mut` is returning Null, the transaction list is empty and we need a new transaction OR ignore the data and move forward since it does not belong to any legit state?

Reply:
More likely ignore the data and move forward, I would say, because the idea is that if a new transaction was needed, it would have been created in the previous step. So you suggest that I return `AppLayerResult::ok()`, here, instead?